src/hotspot/share/memory/metadataFactory.hpp

```diff
@@ -61,15 +63,20 @@ class MetadataFactory : AllStatic {
 
   // Deallocation method for metadata
   template <class T>
-  static void free_metadata(ClassLoaderData* loader_data, T md) {
+  static void free_metadata(ClassLoaderData* loader_data, T md, bool invoke_destructor = false) {
     if (md != NULL) {
       assert(loader_data != NULL, "shouldn't pass null");
       int size = md->size();
       // Call metadata's deallocate function which will call deallocate fields
       assert(!md->on_stack(), "can't deallocate things on stack");
       assert(!md->is_shared(), "cannot deallocate if in shared spaces");
       md->deallocate_contents(loader_data);
-      loader_data->metaspace_non_null()->deallocate((MetaWord*)md, size, md->is_klass());
+      bool is_klass = md->is_klass();
+      if (invoke_destructor) {
+        using U = typename RemoveCV<typename RemovePointer<T>::type>::type;
+        md->~U();
```

Comment:
I was actually thinking of a template specialization for MethodData, but I guess this works. Lets see what others think.

Reply:
I don't think you can do this here.  The preceding deallocate_contents() function will deallocate the metadata so this might crash.  It's like calling the destructor after the delete operator.
Maybe before it the deallocate contents call?
Specializing for MethodData might work if you add code to tell InstanceKlass::deallocate_contents() not to call release_C_heap_structures on methods.

Reply:
Also don't you have to free this?
    FailedSpeculation::free_failed_speculations(method_data()->get_failed_speculations_address());